hashing/src/indexed_merkle_proof.rs

```diff
@@ -0,0 +1,525 @@
+        assert_eq!(
+            out_of_bounds_indexed_merkle_proof.verify(),
+            Err(error::MerkleVerificationError::IndexOutOfBounds {
+                count: 0,
+                index: 23
+            })
+        )
+    }
+
+    #[test]
+    fn deep_proof_doesnt_kill_stack() {
+        const PROOF_LENGTH: usize = 63;
+        let indexed_merkle_proof = IndexedMerkleProof {
+            index: 42,
+            count: 1 << (PROOF_LENGTH - 1),
+            merkle_proof: vec![Digest([0u8; Digest::LENGTH]); PROOF_LENGTH],
+        };
+        let _hash = indexed_merkle_proof.root_hash();
+    }
+
+    #[test]
+    fn empty_proof() {
+        let empty_merkle_root = Digest::hash_merkle_tree(vec![]);
+        assert_eq!(
+            empty_merkle_root,
+            Digest::hash_pair(0u64.to_le_bytes(), Digest::SENTINEL_MERKLE_TREE)
+        );
+        let indexed_merkle_proof = IndexedMerkleProof {
+            index: 0,
+            count: 0,
+            merkle_proof: vec![],
+        };
+        assert_eq!(indexed_merkle_proof.verify(), Ok(()));
+        assert_eq!(indexed_merkle_proof.root_hash(), empty_merkle_root);
+    }
+
+    #[proptest]
+    fn expected_proof_length_le_65(index: u64, count: u64) {
+        let indexed_merkle_proof = IndexedMerkleProof {
+            index,
+            count,
+            merkle_proof: vec![],
+        };
+        prop_assert!(indexed_merkle_proof.compute_expected_proof_length() <= 65);
+    }
+
+    fn reference_root_from_proof(index: u64, count: u64, proof: &[Digest]) -> Digest {
+        fn compute_raw_root_from_proof(index: u64, leaf_count: u64, proof: &[Digest]) -> Digest {
+            if leaf_count == 0 {
+                return Digest::SENTINEL_MERKLE_TREE;
+            }
+            if leaf_count == 1 {
+                return proof[0];
+            }
+            let half = 1u64 << (63 - (leaf_count - 1).leading_zeros());
+            let last = proof.len() - 1;
+            if index < half {
+                let left = compute_raw_root_from_proof(index, half, &proof[..last]);
+                Digest::hash_pair(&left, &proof[last])
+            } else {
+                let right =
+                    compute_raw_root_from_proof(index - half, leaf_count - half, &proof[..last]);
+                Digest::hash_pair(&proof[last], &right)
+            }
+        }
+
+        let raw_root = compute_raw_root_from_proof(index, count, proof);
+        Digest::hash_pair(count.to_le_bytes(), raw_root)
+    }
+
+    /// Construct an `IndexedMerkleProof` with a proof of zero digests.
+    fn test_indexed_merkle_proof(index: u64, count: u64) -> IndexedMerkleProof {
+        let mut indexed_merkle_proof = IndexedMerkleProof {
+            index,
+            count,
+            merkle_proof: vec![],
+        };
+        let expected_proof_length = indexed_merkle_proof.compute_expected_proof_length();
+        indexed_merkle_proof.merkle_proof = std::iter::repeat_with(|| Digest([0u8; 32]))
```

Comment:
In these tests I'd also use different hashes. If all hashes are the same, discrepancies in the implementations could remain undetected, e.g. using the hashes in the wrong order.

Reply:
Good catch.
Random digest is used now.